proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2296,478 @@ crunches doIPCTransfer, possibleSwitchTo
+   apply (case_tac "reply_tcb r"; clarsimp)
+  apply (rule BlockedOnReply_sym_ref1; clarsimp simp: pred_tcb_at_def obj_at_def)
+  done
+
+lemma no_replySC_valid_replies'_sc_asrt:
+  "replySCs_of s r = None \<Longrightarrow> valid_replies'_sc_asrt r s"
+  unfolding valid_replies'_sc_asrt_def
+  by (simp)
+
+crunches doIPCTransfer
+  for replySCs_of[wp]: "\<lambda>s. P (replySCs_of s)"
+  (wp: crunch_wps simp: crunch_simps)
+
+(* FIXME RT: replace get_ep_valid_ep with this better version *)
+lemma get_ep_valid_ep'[wp]:
+  "\<lbrace> valid_objs and ep_at ep \<rbrace>
+   get_endpoint ep
+   \<lbrace> valid_ep \<rbrace>"
+  by (wpsimp simp: ep_at_def2 valid_ep_def2 simp_del: valid_simple_obj_def wp_del: get_ep_valid_ep)
+
+definition receive_reply :: "Structures_A.thread_state \<Rightarrow> obj_ref option" where
+  "receive_reply st \<equiv> case st of
+           (Structures_A.thread_state.BlockedOnReceive ep reply_opt pl) \<Rightarrow> reply_opt"
+
+lemma receive_reply_sel:
+  "receive_reply (Structures_A.thread_state.BlockedOnReceive ep reply_opt pl) = reply_opt"
+  by (simp add: receive_reply_def)
+
+lemma BlockedOnReceive_reply_tcb_reply_at:
+  "\<exists>epptr pl. st_tcb_at ((=) (Structures_A.BlockedOnReceive epptr (Some rptr) pl)) tptr s
+   \<Longrightarrow> sym_refs (state_refs_of s) \<Longrightarrow> valid_objs s
+   \<Longrightarrow> reply_tcb_reply_at (\<lambda>x. x = Some tptr) rptr s"
+  apply (subgoal_tac "reply_at rptr s")
+  apply (clarsimp simp: pred_tcb_at_eq_commute)
+  apply (subgoal_tac "(tptr, ReplyTCB) \<in> (state_refs_of s) rptr")
+  apply (clarsimp simp: state_refs_of_def  get_refs_def reply_tcb_reply_at_def
+                        pred_tcb_at_def obj_at_def refs_of_def is_reply
+                 split: option.splits)
+  apply (erule sym_refsE)
+  apply (clarsimp simp: state_refs_of_def pred_tcb_at_def obj_at_def)
+  apply (clarsimp dest!: st_tcb_at_valid_st2 simp: valid_tcb_state_def)
+  done
+
+lemma sendIPC_corres:
+(* call is only true if called in handleSyscall SysCall, which is always blocking. *)
   assumes "call \<longrightarrow> bl"
   shows
   "corres dc (all_invs_but_fault_tcbs and fault_tcbs_valid_states_except_set {t} and valid_list
-                                      and valid_sched_except_blocked_except_released_ipc_qs
-                                      and st_tcb_at active t and ep_at ep
-                                      and ex_nonz_cap_to t and scheduler_act_not t
-                                      and (\<lambda>s. cd \<longrightarrow> bound_sc_tcb_at (\<lambda>a. \<exists>y. a = Some y) t s))
+              and valid_sched_action and ep_at ep and ex_nonz_cap_to t and tcb_at t
+              and scheduler_act_not t and (\<lambda>s. cd \<longrightarrow> bound_sc_tcb_at (\<lambda>a. \<exists>y. a = Some y) t s))
              (invs' and sch_act_not t and tcb_at' t and ep_at' ep)
              (send_ipc bl call bg cg cgr cd t ep) (sendIPC bl call bg cg cgr cd t ep)"
-  sorry
+proof -
```

Comment:
Is there a reason to start this off as an Isar proof?

Reply:
Nope, only there because I was reusing old proofs to make this one. Will change it.